nomad/event_endpoint.go

```diff
@@ -0,0 +1,211 @@
+package nomad
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/hashicorp/go-msgpack/codec"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/hashicorp/nomad/nomad/stream"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type Event struct {
+	srv *Server
+}
+
+func (e *Event) register() {
+	e.srv.streamingRpcs.Register("Event.Stream", e.stream)
+}
+
+func (e *Event) stream(conn io.ReadWriteCloser) {
+	defer conn.Close()
+
+	var args structs.EventStreamRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&args); err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// forward to appropriate region
+	if args.Region != e.srv.config.Region {
+		err := e.forwardStreamingRPC(args.Region, "Event.Stream", args, conn)
+		if err != nil {
+			handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		}
+		return
+	}
+
+	// ACL check
+	// TODO(drew) ACL checks need to be per topic
+	// All Events        Management
+	// System Events     Management
+	// Node Events       NamespaceCapabilityReadEvents
+	// Job/Alloc Events  NamespaceCapabilityReadEvents
+	if aclObj, err := e.srv.ResolveToken(args.AuthToken); err != nil {
+		handleJsonResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil && !aclObj.IsManagement() {
+		handleJsonResultError(structs.ErrPermissionDenied, helper.Int64ToPtr(403), encoder)
+		return
+	}
+
+	// authToken is passed to the subscribe request so the event stream
+	// can handle closing a subscription if the authToken expires.
+	// If ACLs are disabled, a random token is generated and it will
+	// never be closed due to expiry.
+	authToken := args.AuthToken
+	if authToken == "" {
+		authToken = uuid.Generate()
+	}
+	subReq := &stream.SubscribeRequest{
+		Token:     authToken,
+		Topics:    args.Topics,
+		Index:     uint64(args.Index),
+		Namespace: args.Namespace,
+	}
+	publisher, err := e.srv.State().EventPublisher()
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// start subscription to publisher
+	subscription, err := publisher.Subscribe(subReq)
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+	defer subscription.Unsubscribe()
+
+	ndJsonCh := make(chan *structs.NDJson)
+	errCh := make(chan error)
+
+	jsonStream := stream.NewNDJsonStream(ndJsonCh, 30*time.Second)
+	jsonStream.Run(ctx)
+
+	// goroutine to detect remote side closing
+	go func() {
+		if _, err := conn.Read(nil); err != nil {
+			// One end of the pipe explicitly closed, exit
+			cancel()
+			return
+		}
+		select {
+		case <-ctx.Done():
+			return
+		}
```

Comment:
Not sure I follow the logic here - a couple of questions:

* Will `Read()` ever return a non-error? Specially with passing `nil` bytes, I assume it'll always error.
* What's the value of waiting for ctx.Done()?  It looks like the goroutine will exit regardless?  Is this meant to be inside a loop?

Reply:
Yeah I think this  wasn't properly implemented... in other streaming RPCs we have a goroutine like so. Its not entirely clear to me why its needed

```
	// goroutine to detect remote side closing
	go func() {
		for {
			if _, err := conn.Read(nil); err != nil {
				if err == io.EOF || err == io.ErrClosedPipe {
					// One end of the pipe was explicitly closed, exit cleanly
					cancel()
					return
				}
				select {
				case errCh <- err:
				case <-ctx.Done():
					return
				}
			}
		}
	}()

```